File: backend/app/api/memory.py
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from app.services.memory import MemoryScope, MemoryService

router = APIRouter(prefix="/memory", tags=["memory"])
memory_service = MemoryService()


class MemoryFactDesc(BaseModel):
    id: int
    content: str
    tags: List[str] = Field(default_factory=list)
    created_at: int


class MemoryCandidateDesc(BaseModel):
    id: int
    content: str
    reason: str
    status: str
    created_at: int


class MemorySummaryDesc(BaseModel):
    id: int
    content: str
    created_at: int
    session_id: str


class MemoryFactListResponse(BaseModel):
    facts: List[MemoryFactDesc] = Field(default_factory=list)


class MemoryCandidateListResponse(BaseModel):
    candidates: List[MemoryCandidateDesc] = Field(default_factory=list)


class MemorySummaryListResponse(BaseModel):
    summaries: List[MemorySummaryDesc] = Field(default_factory=list)


class MemoryExportResponse(BaseModel):
    facts: List[Dict[str, Any]] = Field(default_factory=list)
    summaries: List[Dict[str, Any]] = Field(default_factory=list)


class MemoryImportRequest(BaseModel):
    facts: List[Dict[str, Any]] = Field(default_factory=list)
    summaries: List[Dict[str, Any]] = Field(default_factory=list)


class MemoryImportResponse(BaseModel):
    facts: int
    summaries: int


class MemoryActionResponse(BaseModel):
    ok: bool


class MemoryCandidateActionResponse(BaseModel):
    ok: bool
    fact: Optional[MemoryFactDesc] = None


@router.get("/facts", response_model=MemoryFactListResponse)
async def list_memory_facts(
    user_id: str = Query(default="default"),
    profile_id: str = Query(default="default"),
    limit: int = Query(default=50, ge=1, le=500),
) -> MemoryFactListResponse:
    scope = _build_scope(user_id=user_id, profile_id=profile_id)
    facts = memory_service.list_facts(scope, limit=limit)
    return MemoryFactListResponse(
        facts=[
            MemoryFactDesc(
                id=fact.id,
                content=fact.content,
                tags=fact.tags,
                created_at=fact.created_at,
            )
            for fact in facts
        ]
    )


@router.delete("/facts/{fact_id}", response_model=MemoryActionResponse)
async def delete_memory_fact(
    fact_id: int,
    user_id: str = Query(default="default"),
    profile_id: str = Query(default="default"),
) -> MemoryActionResponse:
    scope = _build_scope(user_id=user_id, profile_id=profile_id)
    ok = memory_service.delete_fact(scope, fact_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Memory fact not found")
    return MemoryActionResponse(ok=True)


@router.get("/candidates", response_model=MemoryCandidateListResponse)
async def list_memory_candidates(
    user_id: str = Query(default="default"),
    profile_id: str = Query(default="default"),
    status: str = Query(default="pending"),
    limit: int = Query(default=50, ge=1, le=500),
) -> MemoryCandidateListResponse:
    scope = _build_scope(user_id=user_id, profile_id=profile_id)
    candidates = memory_service.list_candidates(scope, status=status, limit=limit)
    return MemoryCandidateListResponse(
        candidates=[
            MemoryCandidateDesc(
                id=candidate.id,
                content=candidate.content,
                reason=candidate.reason,
                status=candidate.status,
                created_at=candidate.created_at,
            )
            for candidate in candidates
        ]
    )


@router.get("/summaries", response_model=MemorySummaryListResponse)
async def list_memory_summaries(
    user_id: str = Query(default="default"),
    profile_id: str = Query(default="default"),
    limit: int = Query(default=50, ge=1, le=500),
) -> MemorySummaryListResponse:
    scope = _build_scope(user_id=user_id, profile_id=profile_id)
    summaries = memory_service.list_summaries(scope, limit=limit)
    return MemorySummaryListResponse(
        summaries=[
            MemorySummaryDesc(
                id=summary.id,
                content=summary.content,
                created_at=summary.created_at,
                session_id=summary.session_id,
            )
            for summary in summaries
        ]
    )


@router.delete("/summaries/{summary_id}", response_model=MemoryActionResponse)
async def delete_memory_summary(
    summary_id: int,
    user_id: str = Query(default="default"),
    profile_id: str = Query(default="default"),
) -> MemoryActionResponse:
    scope = _build_scope(user_id=user_id, profile_id=profile_id)
    ok = memory_service.delete_summary(scope, summary_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Memory summary not found")
    return MemoryActionResponse(ok=True)


@router.post("/candidates/{candidate_id}/accept", response_model=MemoryCandidateActionResponse)
async def accept_memory_candidate(
    candidate_id: int,
    user_id: str = Query(default="default"),
    profile_id: str = Query(default="default"),
) -> MemoryCandidateActionResponse:
    scope = _build_scope(user_id=user_id, profile_id=profile_id)
    fact = memory_service.accept_candidate(scope, candidate_id)
    if not fact:
        raise HTTPException(status_code=404, detail="Memory candidate not found")
    return MemoryCandidateActionResponse(
        ok=True,
        fact=MemoryFactDesc(
            id=fact.id,
            content=fact.content,
            tags=fact.tags,
            created_at=fact.created_at,
        ),
    )


@router.post("/candidates/{candidate_id}/reject", response_model=MemoryActionResponse)
async def reject_memory_candidate(
    candidate_id: int,
    user_id: str = Query(default="default"),
    profile_id: str = Query(default="default"),
) -> MemoryActionResponse:
    scope = _build_scope(user_id=user_id, profile_id=profile_id)
    ok = memory_service.reject_candidate(scope, candidate_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Memory candidate not found")
    return MemoryActionResponse(ok=True)


@router.get("/export", response_model=MemoryExportResponse)
async def export_memory(
    user_id: str = Query(default="default"),
    profile_id: str = Query(default="default"),
    facts_limit: int = Query(default=200, ge=1, le=2000),
    summaries_limit: int = Query(default=200, ge=1, le=2000),
) -> MemoryExportResponse:
    scope = _build_scope(user_id=user_id, profile_id=profile_id)
    payload = memory_service.export_data(
        scope,
        facts_limit=facts_limit,
        summaries_limit=summaries_limit,
    )
    return MemoryExportResponse(**payload)


@router.post("/import", response_model=MemoryImportResponse)
async def import_memory(
    request: MemoryImportRequest,
    user_id: str = Query(default="default"),
    profile_id: str = Query(default="default"),
) -> MemoryImportResponse:
    scope = _build_scope(user_id=user_id, profile_id=profile_id)
    stats = memory_service.import_data(scope, facts=request.facts, summaries=request.summaries)
    return MemoryImportResponse(**stats)


def _build_scope(*, user_id: str, profile_id: str) -> MemoryScope:
    return MemoryScope(session_id="default", user_id=user_id or "default", profile_id=profile_id or "default")

File: frontend/packages/app-core/src/services/memory.ts
import { appConfig } from "../config";

export type MemoryFact = {
  id: number;
  content: string;
  tags: string[];
  created_at: number;
};

export type MemoryCandidate = {
  id: number;
  content: string;
  reason: string;
  status: string;
  created_at: number;
};

export type MemorySummary = {
  id: number;
  content: string;
  created_at: number;
  session_id: string;
};

export type MemoryExportPayload = {
  facts: Array<{ content: string; tags?: string[]; created_at?: number }>;
  summaries: Array<{ content: string; created_at?: number; session_id?: string }>;
};

const proxyBaseUrl = appConfig.providers.proxyUrl?.trim();
const apiBaseUrl = appConfig.providers.apiBaseUrl?.trim();

function resolveApiBaseUrl() {
  if (proxyBaseUrl) {
    return proxyBaseUrl.replace(/\/$/, "");
  }
  if (apiBaseUrl) {
    return apiBaseUrl.replace(/\/$/, "");
  }
  return "";
}

async function request<T>(path: string, options?: RequestInit) {
  const baseUrl = resolveApiBaseUrl();
  const response = await fetch(`${baseUrl}${path}`, options);
  if (!response.ok) {
    const text = await response.text().catch(() => "");
    const suffix = text ? ` ${text}` : "";
    throw new Error(`Memory API error: ${response.status}${suffix}`);
  }
  try {
    return (await response.json()) as T;
  } catch {
    throw new Error("Memory API returned non-JSON response. Check API base URL.");
  }
}

export async function listMemoryFacts(params: {
  userId: string;
  profileId: string;
  limit?: number;
}): Promise<MemoryFact[]> {
  const search = new URLSearchParams({
    user_id: params.userId,
    profile_id: params.profileId,
  });
  if (params.limit) {
    search.set("limit", String(params.limit));
  }
  const result = await request<{ facts?: MemoryFact[] }>(`/api/memory/facts?${search.toString()}`);
  return result.facts ?? [];
}

export async function deleteMemoryFact(params: {
  factId: number;
  userId: string;
  profileId: string;
}) {
  const search = new URLSearchParams({
    user_id: params.userId,
    profile_id: params.profileId,
  });
  return await request<{ ok: boolean }>(
    `/api/memory/facts/${params.factId}?${search.toString()}`,
    { method: "DELETE" }
  );
}

export async function listMemoryCandidates(params: {
  userId: string;
  profileId: string;
  status?: string;
  limit?: number;
}): Promise<MemoryCandidate[]> {
  const search = new URLSearchParams({
    user_id: params.userId,
    profile_id: params.profileId,
    status: params.status || "pending",
  });
  if (params.limit) {
    search.set("limit", String(params.limit));
  }
  const result = await request<{ candidates?: MemoryCandidate[] }>(
    `/api/memory/candidates?${search.toString()}`
  );
  return result.candidates ?? [];
}

export async function listMemorySummaries(params: {
  userId: string;
  profileId: string;
  limit?: number;
}): Promise<MemorySummary[]> {
  const search = new URLSearchParams({
    user_id: params.userId,
    profile_id: params.profileId,
  });
  if (params.limit) {
    search.set("limit", String(params.limit));
  }
  const result = await request<{ summaries?: MemorySummary[] }>(
    `/api/memory/summaries?${search.toString()}`
  );
  return result.summaries ?? [];
}

export async function deleteMemorySummary(params: {
  summaryId: number;
  userId: string;
  profileId: string;
}) {
  const search = new URLSearchParams({
    user_id: params.userId,
    profile_id: params.profileId,
  });
  return await request<{ ok: boolean }>(
    `/api/memory/summaries/${params.summaryId}?${search.toString()}`,
    { method: "DELETE" }
  );
}

export async function acceptMemoryCandidate(params: {
  candidateId: number;
  userId: string;
  profileId: string;
}): Promise<MemoryFact | null> {
  const search = new URLSearchParams({
    user_id: params.userId,
    profile_id: params.profileId,
  });
  const result = await request<{ ok: boolean; fact?: MemoryFact }>(
    `/api/memory/candidates/${params.candidateId}/accept?${search.toString()}`,
    { method: "POST" }
  );
  return result.fact ?? null;
}

export async function rejectMemoryCandidate(params: {
  candidateId: number;
  userId: string;
  profileId: string;
}) {
  const search = new URLSearchParams({
    user_id: params.userId,
    profile_id: params.profileId,
  });
  return await request<{ ok: boolean }>(
    `/api/memory/candidates/${params.candidateId}/reject?${search.toString()}`,
    { method: "POST" }
  );
}

export async function exportMemory(params: {
  userId: string;
  profileId: string;
}): Promise<MemoryExportPayload> {
  const search = new URLSearchParams({
    user_id: params.userId,
    profile_id: params.profileId,
  });
  return await request<MemoryExportPayload>(`/api/memory/export?${search.toString()}`);
}

export async function importMemory(params: {
  userId: string;
  profileId: string;
  payload: MemoryExportPayload;
}): Promise<{ facts: number; summaries: number }> {
  const search = new URLSearchParams({
    user_id: params.userId,
    profile_id: params.profileId,
  });
  return await request<{ facts: number; summaries: number }>(
    `/api/memory/import?${search.toString()}`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(params.payload),
    }
  );
}
